agent.go

```diff
@@ -278,7 +278,7 @@ func NewAgent(config *AgentConfig) (*Agent, error) { //nolint:gocognit
 
 	a := &Agent{
 		chanTask:          make(chan task),
-		chanState:         make(chan ConnectionState),
+		chanState:         make(chan ConnectionState, 4),
```

Comment:
Why 4 ?
Won't this deadlock if we get unlucky and there is 5 events queued up.

Reply:
We only need this to close the agent correctly. The current behaviour allows agent closing in the callback which with a channel capacity of 0 would lead to deadlock. 

I will see what the right way to handle this is.

Reply:
Considering that we can go Connection to Disconnected to Connected many many times, I don't think a limit to the channel makes sense. Especially considering that we don't want to block the state updating go routine. So I've introduced a queue here which runs the callbacks in a separate goroutine